core/mempool/src/lib.rs

```diff
@@ -23,7 +24,9 @@ const POISONED_LOCK_ERR: &str = "The lock was poisoned.";
 /// Mempool that stores transactions and receipts for a chain
 pub struct Pool {
     signer: Arc<BlockSigner>,
-    transactions: RwLock<HashSet<SignedTransaction>>,
+    // transactions need to be grouped by account, and within each account,
+    // ordered by nonce so that runtime will not ignore some of the transactions
+    transactions: RwLock<HashMap<AccountId, BTreeMap<u64, SignedTransaction>>>,
```

Comment:
Why `BTreeMap` and not `HashMap`?

Reply:
Because the values will be ordered by key so that we don't need to sort them again when we get the values out. Complexity wise I think it's the same. Just makes code easier to write

Reply:
`HashMap` is faster: https://users.rust-lang.org/t/hashmap-performance/6476/1
Deleting element from `HashMap` takes O(1) because it is a hash-map.

Reply:
But we have to do sorting if we use hashmap. That's what I meant by complexity is the same

Reply:
But I understand why we want to use ordered container. It is toally valid.